Feature/jpa liquibase soft delete

Daedrus1 opened the pull request.

Daedrus1 committed:
Add Liquibase dependency

Daedrus1 committed:
"refactor: migrate BookRepository to JpaRepository"

Daedrus1 committed:
"refactor: migrate BookRepository to JpaRepository"

Daedrus1 committed:
soft delete add

Elena-Bruyako commented on src/main/java/mate/academy/springbootweb/controller/BookController.java:
```diff
@@ -0,0 +1,50 @@
+        return service.update(id, dto);
+    }
+
+    // DELETE: /api/books/{id}
```
remove comment

Elena-Bruyako commented on src/main/java/mate/academy/springbootweb/model/Book.java:
```diff
@@ -0,0 +1,44 @@
+@Getter
+@Setter
+@SQLDelete(sql = "UPDATE books SET deleted = true WHERE id = ?")
+@Where(clause = "deleted = false")
```
`@Where` is deprecated
use SQLRestriction instead

Elena-Bruyako commented on src/main/java/mate/academy/springbootweb/model/Book.java:
```diff
@@ -0,0 +1,44 @@
+    @Column(length = 512)
+    private String coverImage;
+    @Column(nullable = false)
+    private boolean deleted = false;
```
```suggestion
    private boolean isDeleted = false;
```

Elena-Bruyako replied:
fix comment

Elena-Bruyako commented on src/main/java/mate/academy/springbootweb/service/BookServiceImpl.java:
```diff
@@ -0,0 +1,67 @@
+        Book current = repository.findById(id)
+                .orElseThrow(() -> new EntityNotFoundException("Book " + id + " not found"));
+
+        // Полный PUT: перезаписываем все поля значениями из DTO
```
remove comment

Elena-Bruyako commented on src/main/java/mate/academy/springbootweb/service/BookServiceImpl.java:
```diff
@@ -0,0 +1,67 @@
+                .orElseThrow(() -> new EntityNotFoundException("Book " + id + " not found"));
+
+        // Полный PUT: перезаписываем все поля значениями из DTO
+        current.setTitle(update.getTitle());
```
use mapper to update book

Elena-Bruyako commented on src/main/java/mate/academy/springbootweb/service/BookServiceImpl.java:
```diff
@@ -0,0 +1,67 @@
+        current.setCoverImage(update.getCoverImage());
+
+        Book saved = repository.save(current);
+        return mapper.toDto(saved);
```
Book is reference type
```suggestion
        repository.save(current);
        return mapper.toDto(current);
```

Elena-Bruyako commented on src/main/java/mate/academy/springbootweb/service/BookServiceImpl.java:
```diff
@@ -0,0 +1,67 @@
+
+    @Override
+    public BookDto update(Long id, BookDto update) {
+        Book current = repository.findById(id)
```
current - is bad naming

Elena-Bruyako commented on src/main/resources/db/changelog/001-create-books.yaml:
```diff
@@ -0,0 +1,41 @@
+                    nullable: false
+              - column:
+                  name: description
+                  type: CLOB
```
```suggestion
                  type: VARCHAR(255)
```

Elena-Bruyako commented on src/main/resources/db/changelog/001-create-books.yaml:
```diff
@@ -0,0 +1,41 @@
+                  type: CLOB
+              - column:
+                  name: cover_image
+                  type: VARCHAR(1024)
```
```suggestion
                  type: VARCHAR(255)
```

Elena-Bruyako reviewed (changes requested):
see comments

Daedrus1 committed:
- Replaced manual field setters in BookServiceImpl.update() with BookMapper.updateEntityFromDto()
- Renamed variable `current` to `existingBook` for clarity
- Replaced deprecated @Where with @SQLRestriction in Book entity for soft delete

Elena-Bruyako commented on src/main/java/mate/academy/springbootweb/controller/BookController.java:
```diff
@@ -0,0 +1,49 @@
+    }
+
+    @PutMapping("/{id}")
+    public BookDto update(@PathVariable Long id, @RequestBody BookDto dto) {
```
Have you checked your solution in Postman?

Elena-Bruyako replied:
```suggestion
    public BookDto update(@PathVariable Long id, @RequestBody CreateBookRequestDto dto) {
```

Elena-Bruyako commented on src/main/java/mate/academy/springbootweb/mapper/BookMapper.java:
```diff
@@ -0,0 +1,25 @@
+
+    @Mapping(target = "id", ignore = true)
+    @Mapping(target = "deleted", ignore = true)
+    void updateEntityFromDto(BookDto dto, @MappingTarget Book target);
```
remove

Elena-Bruyako commented on src/main/java/mate/academy/springbootweb/model/Book.java:
```diff
@@ -0,0 +1,44 @@
+@Getter
+@Setter
+@SQLDelete(sql = "UPDATE books SET deleted = true WHERE id = ?")
+@SQLRestriction("deleted = false")
```
```suggestion
@SQLRestriction("is_deleted = false")
```

Elena-Bruyako commented on src/main/java/mate/academy/springbootweb/model/Book.java:
```diff
@@ -0,0 +1,44 @@
+        ))
+@Getter
+@Setter
+@SQLDelete(sql = "UPDATE books SET deleted = true WHERE id = ?")
```
```suggestion
@SQLDelete(sql = "UPDATE books SET is_deleted = true WHERE id = ?")
```

Elena-Bruyako commented on src/main/java/mate/academy/springbootweb/model/Book.java:
```diff
@@ -0,0 +1,44 @@
+    private BigDecimal price;
+    @Column(columnDefinition = "TEXT")
+    private String description;
+    @Column(length = 512)
```
```suggestion
    private String description;
```

Elena-Bruyako commented on src/main/java/mate/academy/springbootweb/service/BookServiceImpl.java:
```diff
@@ -0,0 +1,61 @@
+        mapper.updateEntityFromDto(update, existingBook);
+
+        Book savedBook = repository.save(existingBook);
+        return mapper.toDto(savedBook);
```
```suggestion
        repository.save(existingBook);
        return mapper.toDto(existingBook);
```

Elena-Bruyako reviewed (changes requested):
see comments

Daedrus1 committed:
- Replaced manual field setters in BookServiceImpl.update() with BookMapper.updateEntityFromDto()
- Renamed variable `current` to `existingBook` for clarity
- Replaced deprecated @Where with @SQLRestriction in Book entity for soft delete

Daedrus1 committed:
- Replaced manual field setters in BookServiceImpl.update() with BookMapper.updateEntityFromDto()
- Renamed variable `current` to `existingBook` for clarity
- Replaced deprecated @Where with @SQLRestriction in Book entity for soft delete

Elena-Bruyako commented on src/main/java/mate/academy/springbootweb/repository/BookRepository.java:
```diff
@@ -0,0 +1,8 @@
+import org.springframework.data.jpa.repository.JpaRepository;
+
+public interface BookRepository extends JpaRepository<Book, Long> {
+    boolean existsByIsbn(String isbn);
```
remove

Elena-Bruyako commented on src/main/java/mate/academy/springbootweb/service/BookServiceImpl.java:
```diff
@@ -0,0 +1,61 @@
+        mapper.updateEntityFromDto(dto, existingBook);
+
+        Book savedBook = repository.save(existingBook);
+        return mapper.toDto(savedBook);
```
```suggestion
        repository.save(existingBook);
        return mapper.toDto(existingBook);
```

Elena-Bruyako commented on src/main/java/mate/academy/springbootweb/service/BookServiceImpl.java:
```diff
@@ -0,0 +1,61 @@
+    }
+
+    @Override
+    public BookDto update(Long id, BookDto dto) {
```
```suggestion
    public BookDto update(Long id, CreateBookRequestDto dto) {
```

Elena-Bruyako commented on src/main/java/mate/academy/springbootweb/dto/UpdateBookRequestDto.java:
```diff
@@ -0,0 +1,10 @@
+package mate.academy.springbootweb.dto;
+
+public class UpdateBookRequestDto {
```
remove

Elena-Bruyako commented on src/main/java/mate/academy/springbootweb/mapper/BookMapper.java:
```diff
@@ -0,0 +1,27 @@
+
+    @Mapping(target = "id", ignore = true)
+    @Mapping(target = "deleted", ignore = true)
+    void updateEntityFromDto(BookDto dto, @MappingTarget Book target);
```
```suggestion
```

Elena-Bruyako commented on src/main/java/mate/academy/springbootweb/mapper/BookMapper.java:
```diff
@@ -0,0 +1,27 @@
+public interface BookMapper {
+    BookDto toDto(Book book);
+
+    BookDto toDto(CreateBookRequestDto dto);
```
```suggestion
```

Elena-Bruyako reviewed (changes requested):
see comments
there are 2 open PRs, merge one of them if it was approved

Daedrus1 committed:
fix

Elena-Bruyako reviewed (changes requested):
there are changes from previous HW, fix it please

Elena-Bruyako commented on src/main/java/mate/academy/springbootweb/model/Book.java:
```diff
@@ -30,9 +34,8 @@ public class Book {
-    private String description;
-    @Column(length = 512)
     private String coverImage;
+    @Column(name = "is_deleted", nullable = false)
```
```suggestion
    @Column(nullable = false)
```

Elena-Bruyako commented on src/main/java/mate/academy/springbootweb/service/BookServiceImpl.java:
```diff
@@ -38,4 +38,23 @@ public BookDto create(CreateBookRequestDto dto) {
+
+        mapper.updateEntityFromDto(dto, entity);
+        Book saved = repository.save(entity);
+        return mapper.toDto(saved);
```
```suggestion
        repository.save(entity);
        return mapper.toDto(entity);
```

Elena-Bruyako commented on src/main/java/mate/academy/springbootweb/service/BookServiceImpl.java:
```diff
@@ -38,4 +38,23 @@ public BookDto create(CreateBookRequestDto dto) {
+
+    @Override
+    public BookDto update(Long id, CreateBookRequestDto dto) {
+        Book entity = repository.findById(id)
```
```suggestion
        Book book = repository.findById(id)
```

Elena-Bruyako reviewed (changes requested):
see comments

Daedrus1 committed:
fix

Elena-Bruyako commented on src/main/java/mate/academy/springbootweb/model/Book.java:
```diff
@@ -30,9 +34,8 @@ public class Book {
-    @Column(length = 512)
     private String coverImage;
+    @Column(nullable = false)
+    private boolean deleted = false;
```
```suggestion
    private boolean isDeleted = false;
```

Elena-Bruyako reviewed (changes requested):
one comment

Daedrus1 committed:
fix
